File: day26/BinaryTraverse.java
import java.util.*;

class TreeNode {
    int val;
    TreeNode left, right;
    TreeNode(int val) { this.val = val; }
}

public class BinaryTraverse {

    public static List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();
        if (root == null) return result;

        Queue<TreeNode> queue = new ArrayDeque<>();
        queue.offer(root);

        while (!queue.isEmpty()) {
            int size = queue.size();
            List<Integer> level = new ArrayList<>();
            while (size-- > 0) {
                TreeNode node = queue.poll();
                level.add(node.val);
                if (node.left != null) queue.offer(node.left);
                if (node.right != null) queue.offer(node.right);
            }
            result.add(level);
        }

        return result;
    }

    public static List<Integer> treeToList(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        if (root == null) return result;

        Queue<TreeNode> queue = new LinkedList<>(); // allows nulls
        queue.offer(root);
        while (!queue.isEmpty()) {
            TreeNode node = queue.poll();
            if (node != null) {
                result.add(node.val);
                queue.offer(node.left);
                queue.offer(node.right);
            } else {
                result.add(null);
            }
        }

        // Trim trailing nulls
        int i = result.size() - 1;
        while (i >= 0 && result.get(i) == null) {
            result.remove(i--);
        }

        return result;
    }

    public static TreeNode generateRandomTree(int n) {
        if (n == 0) return null;

        Random rand = new Random();
        TreeNode[] nodes = new TreeNode[n];
        for (int i = 0; i < n; i++) {
            nodes[i] = new TreeNode(rand.nextInt(100));
        }

        for (int i = 0; i < n; i++) {
            int left = 2 * i + 1;
            int right = 2 * i + 2;
            if (left < n) nodes[i].left = nodes[left];
            if (right < n) nodes[i].right = nodes[right];
        }

        return nodes[0];
    }

    public static void main(String[] args) {
        int n = new Random().nextInt(6) + 5; // 5–10 nodes
        TreeNode root = generateRandomTree(n);

        System.out.println("Generated binary tree with " + n + " nodes:");
        System.out.println("Tree as level-order list:");
        System.out.println(treeToList(root));
        System.out.println("\nLevel order traversal (grouped by level):");
        System.out.println(levelOrder(root));
    }
}

File: day26/binaryTraverse.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <cstdlib>
#include <ctime>

// C++ 11 compliant

struct TreeNode {
    int val;
    TreeNode* left = nullptr;
    TreeNode* right = nullptr;
    TreeNode(int v) : val(v) {}
};

std::vector<std::vector<int>> levelOrder(TreeNode* root) {
    std::vector<std::vector<int>> result;
    if (!root) return result;

    std::queue<TreeNode*> q;
    q.push(root);

    while (!q.empty()) {
        int size = q.size();
        std::vector<int> level;
        while (size--) {
            TreeNode* node = q.front(); q.pop();
            level.push_back(node->val);
            if (node->left) q.push(node->left);
            if (node->right) q.push(node->right);
        }
        result.push_back(level);
    }

    return result;
}

// Using -1 as a sentinel value for null (assuming random values are always 0–99)
std::vector<int> treeToList(TreeNode* root) {
    std::vector<int> result;
    if (!root) return result;

    std::queue<TreeNode*> q;
    q.push(root);

    while (!q.empty()) {
        TreeNode* node = q.front(); q.pop();
        if (node) {
            result.push_back(node->val);
            q.push(node->left);
            q.push(node->right);
        } else {
            result.push_back(-1);  // sentinel for null
        }
    }

    // Trim trailing -1s
    while (!result.empty() && result.back() == -1) {
        result.pop_back();
    }

    return result;
}

TreeNode* generateRandomTree(int n) {
    if (n == 0) return nullptr;

    std::vector<TreeNode*> nodes;
    for (int i = 0; i < n; ++i) {
        nodes.push_back(new TreeNode(rand() % 100));
    }

    for (int i = 0; i < n; ++i) {
        int left = 2 * i + 1;
        int right = 2 * i + 2;
        if (left < n) nodes[i]->left = nodes[left];
        if (right < n) nodes[i]->right = nodes[right];
    }

    return nodes[0];
}

int main() {
    std::srand(static_cast<unsigned>(std::time(nullptr)));
    int n = std::rand() % 6 + 5; // 5–10 nodes
    TreeNode* root = generateRandomTree(n);

    std::cout << "Generated binary tree with " << n << " nodes:\n";
    std::cout << "Tree as level-order list (-1 = null):\n";
    std::vector<int> flat = treeToList(root);
    for (int val : flat) {
        std::cout << (val == -1 ? "null" : std::to_string(val)) << " ";
    }

    std::cout << "\n\nLevel order traversal (grouped by level):\n";
    std::vector<std::vector<int>> levels = levelOrder(root);
    for (const auto& level : levels) {
        std::cout << "[ ";
        for (int val : level) std::cout << val << " ";
        std::cout << "]\n";
    }

    return 0;
}
